scripts/migration/schema-0.2-GPII-4014/createSimulatedGpiiKeys.js

```diff
@@ -0,0 +1,220 @@
+            "revokedReason": null,
+            "timestampCreated": currentTime,
+            "timestampUpdated": null
+        };
+
+        var newPrefsSafe = {
+            "_id": prefsSafeId,
+            "type": "prefsSafe",
+            "schemaVersion": "0.1",
+            "prefsSafeType": "user",
+            "name": gpiiKeyId,
+            "password": null,
+            "email": null,
+            "preferences": prefs,
+            "timestampCreated": currentTime,
+            "timestampUpdated": null
+        };
+        newDocs.push(newGpiiKey);
+        newDocs.push(newPrefsSafe);
+    };
+    return newDocs;
+};
+
+/**
+ * Log how many GPII documents were updated.
+ * @param {String} responseString - Response from the database (ignored)
+ * @param {Object} options - Object containing the set of documents:
+ * @param {Array} options.numOfKeysToCreate - Total number of GPII keys to create.
+ * @param {Array} options.numOfCreatedKeys - Total number of GPII keys that have been created.
+ * @return {Promise} - The resolved value is options.numOfCreatedKeys.
+ */
+gpii.migration.GPII4014.logUpdateDB = function (responseString, options) {
+    var togo = fluid.promise();
+    options.numOfCreatedKeys = options.numOfCreatedKeys + options.numToCreateInThisBatch;
+    if (options.numOfCreatedKeys < options.numOfKeysToCreate) {
+        togo.resolve(options.numOfCreatedKeys);
+    } else {
+        togo.reject({
+            errorCode: "GPII-CREATED-ENOUGH",
+            numOfCreatedKeys: options.numOfCreatedKeys
+        });
+    }
+    return togo;
+};
+
+/**
+ * Configure one batch of creation.
+ * @param {Object} options - Object containing the set of documents:
+ * @param {Array} options.newDocs - The documents to create.
+ * @param {Array} options.numOfKeysToCreate - Total number of GPII keys to create.
+ * @param {Number} options.numOfCreatedKeys - Total number of GPII keys that have been created. This option will be
+ * written by a call to this function.
+ * @return {Promise} - The promise that resolves the creation of this batch.
+ */
+gpii.migration.GPII4014.createOneBatch = function (options) {
+    var numOfNeeded = options.numOfKeysToCreate - options.numOfCreatedKeys;
+    options.numToCreateInThisBatch = numOfNeeded > options.maxDocsInBatchPerRequest ? options.maxDocsInBatchPerRequest : numOfNeeded;
+
+    options.newDocs = gpii.migration.GPII4014.generateKeyData(options.numToCreateInThisBatch);
+
+    var details = {
+        dataToPost: options.newDocs,
+        responseDataHandler: gpii.migration.GPII4014.logUpdateDB
+    };
+    return gpii.dbRequest.configureStep(details, options);
+};
+
+/**
+ * Create GPII keys recursively.
+ * @param {Object} options - Object containing the set of information for GPII key creation.
+ * @param {Promise} togo - The return promise provided by the caller. It will be modified by this function.
+ */
+gpii.migration.GPII4014.createRecursive = function (options, togo) {
```

Comment:
This works, but the signature is a bit nutty. It should return a promise rather than accepting one. Also, we should get rid of the strange abuse of the reject semantic via GPII-CREATED-ENOUGH and simply turn this into a resolve with a value of 0 indicating that no docs remain to be created. Then we can simply bail out of the recursion on receiving this resolution, and the top-level driver can make a bit more sense.